storage/range.go

```diff
@@ -1370,3 +1402,119 @@ func (r *Range) AdminSplit(args *proto.AdminSplitRequest, reply *proto.AdminSpli
 		reply.SetGoError(util.Errorf("split at key %q failed: %s", splitKey, err))
 	}
 }
+
+// InitialState implements the raft.Storage interface.
+func (r *Range) InitialState() (raftpb.HardState, raftpb.ConfState, error) {
+	data, err := r.rm.Engine().Get(
+		proto.EncodedKey(engine.RaftStateKey(uint64(r.Desc.RaftID))))
+	if err != nil {
+		return raftpb.HardState{}, raftpb.ConfState{}, err
+	}
+	var hs raftpb.HardState
+	err = gogoproto.Unmarshal(data, &hs)
+	if err != nil {
+		return raftpb.HardState{}, raftpb.ConfState{}, err
+	}
+
+	// TODO(bdarnell): load ConfState if we have a previous snapshot.
+	return hs, raftpb.ConfState{}, nil
+}
+
+// Entries implements the raft.Storage interface.
+// TODO(bdarnell): consider caching for recent entries.
+func (r *Range) Entries(lo, hi uint64) ([]raftpb.Entry, error) {
+	var ents []raftpb.Entry
+	// Scan over the log (which is stored backwards) to find the
+	// requested entries.  Reversing [lo, hi) gives us (hi, lo]; since
+	// Iterate is inclusive in the other direction we must increment both the
+	// start and end keys.
+	err := r.rm.Engine().Iterate(
```

Comment:
Again, you want to use engine.MVCCScan() here instead and not cast your keys to proto.EncodedKey.